src/modules/dashboard/components/Dialogs/GnosisControlSafeDialog/SafeTransactionPreview.tsx

```diff
@@ -119,84 +138,173 @@ interface Props {
   values: FormValues;
 }
 
-const SafeTransactionPreview = ({ colony: { tokens }, values }: Props) => {
-  const transactionDetails = useMemo(
-    () =>
-      values.transactions[0].transactionType === ''
-        ? []
-        : transactionTypeFieldsMap[values.transactions[0].transactionType],
-    [values],
+const SafeTransactionPreview = ({ colony, values }: Props) => {
+  const [transactionTabStatus, setTransactionTabStatus] = useState(
+    Array(values.transactions.length).fill(false),
   );
 
-  const token = useMemo(
-    () =>
-      tokens.find(
-        (item) => item.address === values.transactions[0].tokenAddress,
-      ),
-    [values, tokens],
+  const tokens = useMemo(() => {
+    return values.transactions.map((transaction) => {
+      const token = colony.tokens.find(
+        (item) => item.address === transaction.tokenAddress,
+      );
+      return token;
+    });
+  }, [values, colony.tokens]);
+
+  const handleTabToggle = useCallback(
+    (newIndex: number) => {
+      const newTransactionTabs = transactionTabStatus.map((tab, index) =>
+        index === newIndex ? !tab : tab,
+      );
+      setTransactionTabStatus([...newTransactionTabs]);
+    },
+    [transactionTabStatus, setTransactionTabStatus],
+  );
+  const getTransactionTypeLabel = useCallback(
+    (transactionTypeValue: string) => {
+      const transactionType = transactionOptions.find(
+        (transaction) => transaction.value === transactionTypeValue,
+      );
+      return transactionType ? (
+        <FormattedMessage {...transactionType.label} />
+      ) : null;
+    },
+    [],
+  );
```

Comment:
Same question here re memoisation. Putting a console.log in the function shows it gets re-run every render, so not sure why we'd want to memoise it. The `Heading` that consumes this function also gets recreated every render because the `values.transactions.map` on line 195 is re-run every render, creating new props for that component.